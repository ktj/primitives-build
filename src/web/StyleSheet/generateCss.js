const normalizeValue = require('./normalizeValue');
const hyphenate = require('./hyphenate');
const mapKeyValue = require('../util/mapKeyValue');
const transformToWebStyle = require('./transformToWebStyle');

/**
 * Generates valid CSS rule bodies (sans selector) from a JS object
 *
 * Example:
 * ```
 * generateCss({ width: 20, color: 'blue' });
 * //=> 'width:20px;color:blue;'
 * ```
 *
 * @param style
 */
const generateCss = style => mapKeyValue(transformToWebStyle(style), (key, val) => {
  const name = hyphenate(key);
  const value = normalizeValue(key, val);
  if (Array.isArray(val)) {
    return val.map(v => `${name}:${v};`).join('');
  }
  return `${name}:${value};`;
}).join('');

module.exports = generateCss;
