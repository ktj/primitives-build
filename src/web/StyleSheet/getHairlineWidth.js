// Modeled after: http://dieulot.net/css-retina-hairline

// NOTE:
// This assumes this script runs after <body>. If it is run inside of `<head>`, it will just
// return 1.
module.exports = () => {
  let hairlineWidth = 1;
  if (!global || !global.document || !global.document.body) {
    return hairlineWidth;
  }
  if (global.devicePixelRatio && global.devicePixelRatio >= 2) {
    const testElem = document.createElement('div');
    testElem.style.border = '.5px solid transparent';
    document.body.appendChild(testElem);
    if (testElem.offsetHeight === 1) {
      hairlineWidth = 0.5;
    }
    document.body.removeChild(testElem);
  }
  return hairlineWidth;
};
