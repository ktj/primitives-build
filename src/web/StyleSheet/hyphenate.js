// Taken from:
// https://github.com/necolas/react-native-web/blob/master/src/apis/StyleSheet/hyphenate.js

module.exports = s => s.replace(/([A-Z])/g, '-$1').toLowerCase().replace(/^ms-/, '-ms-');
