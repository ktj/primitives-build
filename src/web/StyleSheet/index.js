const getHairlineWidth = require('./getHairlineWidth');
const transformToWebStyle = require('./transformToWebStyle');
const { reset, registerStyle, getStyle, getClassNames } = require('./registry');
const init = require('./initialStyles');
const injector = require('./injector');

// TODO:
// 3. ensure that multiple-bundled react-primitives will work...
// 6. use invariant and add a lot of validation + runtime checks
// 7. introduce a "strict mode" which only allows RN-compatible API (not sure about this any more)
// 12. how to handle custom font families??

const hasOwnProperty = Object.prototype.hasOwnProperty;

const create = styles => {
  const result = {};
  Object.keys(styles).forEach(key => {
    result[key] = registerStyle(key, styles[key]);
  });
  return result;
};

const mergeTransforms = (a, b) => {
  if (!a || a.length === 0) return b; // in this case, a has nothing to contribute.
  const result = [];
  const transformsInA = a.reduce((hash, t) => {
    const key = Object.keys(t)[0];
    result.push(t);
    hash[key] = result.length - 1;
    return hash;
  }, {});
  b.forEach(t => {
    const key = Object.keys(t)[0];
    const index = transformsInA[key];
    if (index !== undefined) {
      result[index] = t;
    } else {
      result.push(t);
    }
  });
  return result;
};

// merge two style hashes together. Sort of like `Object.assign`, but is aware of `transform` as a
// special case.
// NOTE: mutates the first argument!
const assignStyle = (a, b) => {
  let key;
  for (key in b) {
    if (hasOwnProperty.call(b, key)) {
      switch (key) {
        case 'transform':
          a[key] = mergeTransforms(a[key], b[key]);
          break;
        default:
          /* eslint no-param-reassign: 0 */
          a[key] = b[key];
          break;
      }
    }
  }
  return a;
};

const flattenStyle = (input, expandRegisteredStyles) => {
  // optimize for this (common) scenario of input not needing flattening
  if (!Array.isArray(input)) {
    if (typeof input === 'number') {
      return expandRegisteredStyles ? getStyle(input) : undefined;
    }
    if (!input) {
      return undefined;
    }
    return input;
  }
  const result = {};
  let hasResult = false;
  const stack = [{ i: 0, array: input }];
  while (stack.length) {
    const iterator = stack.pop();
    let i = iterator.i;
    const array = iterator.array;
    for (; i < array.length; i++) {
      const el = array[i];
      if (Array.isArray(el)) {
        stack.push({ i: i + 1, array });
        stack.push({ i: 0, array: el });
        break;
      }
      if (typeof el === 'number') {
        if (expandRegisteredStyles) {
          hasResult = true;
          assignStyle(result, getStyle(el));
        }
      } else if (el) {
        hasResult = true;
        assignStyle(result, el);
      }
    }
  }
  return hasResult ? result : undefined;
};

const flattenClassNames = (input, flag) => {
  // optimize for this (common) scenario of input not needing flattening
  if (!Array.isArray(input)) {
    if (typeof input === 'number') {
      return getClassNames(input, 0);
    }
    return null;
  }
  let result = '';
  let position = 0;
  let hasSeenObject = false;
  const stack = [{ i: 0, array: input }];
  while (stack.length) {
    const iterator = stack.pop();
    let i = iterator.i;
    const array = iterator.array;
    for (; i < array.length; i++) {
      const el = array[i];
      if (Array.isArray(el)) {
        stack.push({ i: i + 1, array });
        stack.push({ i: 0, array: el });
        break;
      }
      if (typeof el === 'number') {
        result += `${getClassNames(el, position)} `;
        if (hasSeenObject) {
          // TODO(lmr): warn to console in this case to let user know? or do we not care?
          flag.deopt = true;
        }
      } else if (el) {
        hasSeenObject = true;
      }
      position++;
    }
  }
  return result;
};

const resolve = (styles, extraClassName) => {
  const flag = { deopt: false };
  const classes = flattenClassNames(styles, flag);
  const style = flattenStyle(styles, flag.deopt);
  return {
    className: !classes || flag.deopt ? extraClassName : `${extraClassName} ${classes}`,
    style: style ? transformToWebStyle(style) : null,
  };
};

const returnCopy = (original, result) => {
  if (original === result || typeof original === 'number') {
    return { ...result };
  }
  return result;
};

init();

// TODO(lmr):
// figure out a good strategy for SSR here.

const absoluteFillObject = {
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

module.exports = {
  hairlineWidth: getHairlineWidth(),
  absoluteFillObject,
  absoluteFill: registerStyle('absoluteFill', absoluteFillObject),
  create,
  // NOTE:
  // `flatten` is exported separately from `resolve` because it mimics the RN api more closely
  // than `resolve`.
  flatten: style => returnCopy(style, flattenStyle(style, true)) || {},

  // TODO(lmr): should this be an internal API or something that we expose?
  resolve,

  reset,

  getStyleSheetHtml: injector.getStyleSheetHtml,
};
