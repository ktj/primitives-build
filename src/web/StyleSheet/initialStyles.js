const generateCss = require('./generateCss');
const injector = require('./injector');

function init() {
  const initialRules = {
    // https://github.com/facebook/css-layout#default-values
    rp_View: {
      alignItems: 'stretch',
      borderWidth: 0,
      borderStyle: 'solid',
      boxSizing: 'border-box',
      JsDisplay: 'flex',
      display: 'flex',
      flexBasis: 'auto',
      flexDirection: 'column',
      flexShrink: 0,
      margin: 0,
      padding: 0,
      position: 'relative',
      // button and anchor reset
      backgroundColor: 'transparent',
      color: 'inherit',
      font: 'inherit',
      textAlign: 'inherit',
      textDecorationLine: 'none',
      // list reset
      listStyle: 'none',
      // fix flexbox bugs
      minHeight: 0,
      minWidth: 0,
    },
    rp_ViewReset: {
      flexShrink: 0,
    },
    rp_Text: {
      color: 'inherit',
      display: 'inline',
      font: 'inherit',
      margin: 0,
      padding: 0,
      textDecorationLine: 'none',
      wordWrap: 'break-word',
      whiteSpace: 'pre-wrap',
    },
    rp_Image: {
      borderWidth: 0,
      height: 'auto',
      maxHeight: '100%',
      maxWidth: '100%',
      opacity: 0,
    },
  };

  Object.keys(initialRules).forEach(key => {
    const cssBody = generateCss(initialRules[key]);
    const css = `.${key}{${cssBody}}`;
    injector.addRule(key, css);
  });
}

module.exports = init;
