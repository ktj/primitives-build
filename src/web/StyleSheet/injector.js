const asap = require('asap');

const hasOwnProperty = Object.prototype.hasOwnProperty;

// TODO(lmr): to enable multiple bundles, we *could* just try and have registry be a global
let registry = {};
let isDirty = false;
let styleNode = null;

const getStyleText = () => {
  /* eslint prefer-template:0 */
  let result = '\n';
  for (const key in registry) {
    if (hasOwnProperty.call(registry, key)) {
      result += registry[key] + '\n';
    }
  }
  return result;
};

const frame = () => {
  if (!isDirty) return;
  if (!global.document) return;
  const document = global.document;
  isDirty = false;

  // TODO(lmr):
  // in the case of server-side rendering, we might want to include a "checksum" or
  // something to the end of the stylesheet so that we can ensure that we are re-initializing the
  // right one?  I'm not exactly sure how this would work though. The user might have to ensure
  // that there are not more than one "react-primitive" instances running on the page at once.
  if (!styleNode && document.querySelector) {
    styleNode = document.querySelector('style[data-react-primitives-stylesheet]');
  }

  if (!styleNode) {
    // Taken from
    // http://stackoverflow.com/questions/524696/how-to-create-a-style-tag-with-javascript
    const head = document.head || document.getElementsByTagName('head')[0];
    styleNode = document.createElement('style');

    styleNode.type = 'text/css';
    styleNode.setAttribute('data-react-primitives-stylesheet', '');
    head.appendChild(styleNode);
  }

  const css = getStyleText();

  if (styleNode.styleSheet) {
    styleNode.styleSheet.cssText = css;
  } else {
    /* eslint no-cond-assign:0 */
    let last;
    while (last = styleNode.lastChild) {
      styleNode.removeChild(last);
    }
    styleNode.appendChild(document.createTextNode(css));
  }
};

const addRule = (key, rule) => {
  registry[key] = rule;
  if (!isDirty) {
    isDirty = true;
    if (global.document) {
      // NOTE(lmr):
      // we are only calling this in a browser environment. On the server we can just use the
      // `getStyleSheetHtml` API.
      asap(frame);
    }
  }
};

const getStyleSheetHtml = () => `<style data-react-primitives-stylesheet>${getStyleText()}</style>`;

module.exports = {
  addRule,
  getStyleSheetHtml,
  reset: () => { registry = {}; },
};
