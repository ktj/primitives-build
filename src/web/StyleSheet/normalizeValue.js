// Taken from:
// https://github.com/necolas/react-native-web/blob/master/src/apis/StyleSheet/normalizeValue.js
const unitlessNumbers = {
  boxFlex: true,
  boxFlexGroup: true,
  columnCount: true,
  flex: true,
  flexGrow: true,
  flexPositive: true,
  flexShrink: true,
  flexNegative: true,
  fontWeight: true,
  lineClamp: true,
  opacity: true,
  order: true,
  orphans: true,
  widows: true,
  zIndex: true,
  zoom: true,
  // SVG-related
  fillOpacity: true,
  strokeDashoffset: true,
  strokeOpacity: true,
  strokeWidth: true,
  scaleX: true,
  scaleY: true,
  scaleZ: true,
  scale: true,
};

const normalizeValue = (property, value) => {
  if (!unitlessNumbers[property] && typeof value === 'number') {
    return `${value}px`;
  }
  return value;
};

module.exports = normalizeValue;
