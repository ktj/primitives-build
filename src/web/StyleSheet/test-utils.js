const addNewlines = str => str.replace(/([;{,])/g, '$1\n');
const addNewline = str => str.replace(/([{])/g, '\n$1');
const stripWhiteSpace = str => str.replace(/\n\s*/g, '\n').replace(/^\n/, '');
const stripNewlines = str => str.replace(/\n/, '');
const stripHtml = str => str.replace(/(<([^>]+)>)/ig, '');
const formatCss = str => addNewline(addNewlines(stripHtml(str)));

module.exports = {
  formatCss,
  addNewlines,
  stripWhiteSpace,
  stripHtml,
  stripNewlines,
};
