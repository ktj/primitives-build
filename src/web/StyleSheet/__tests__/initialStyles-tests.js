import StyleSheet from '../';
import init from '../initialStyles';
import { formatCss } from '../test-utils';

const flushedCss = () => formatCss(StyleSheet.getStyleSheetHtml());

describe('Core Styles', () => {

  beforeEach(() => StyleSheet.reset());

  it('Initial Styles match', () => {
    init();
    expect(flushedCss()).toMatchSnapshot();
  });
});
