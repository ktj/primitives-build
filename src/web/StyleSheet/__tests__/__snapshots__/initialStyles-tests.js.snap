exports[`Core Styles Initial Styles match 1`] = `
"
.rp_View
{
-webkit-flex-shrink:0px;
-ms-flex-negative:0px;
-moz-box-sizing:border-box;
-webkit-flex-basis:auto;
-webkit-flex-direction:column;
-ms-flex-direction:column;
-webkit-align-items:stretch;
-ms-flex-align:stretch;
-webkit-box-align:stretch;
-ms-preferred-size:auto;
-webkit-box-orient:vertical;
-webkit-box-direction:normal;
color:inherit;
display:-webkit-box;
display:-moz-box;
display:-ms-flexbox;
display:-webkit-flex;
display:flex;
flex-basis:auto;
flex-direction:column;
flex-shrink:0;
font:inherit;
list-style:none;
margin-top:0px;
margin-right:0px;
margin-bottom:0px;
align-items:stretch;
min-height:0px;
min-width:0px;
padding-top:0px;
padding-right:0px;
padding-bottom:0px;
padding-left:0px;
position:relative;
text-align:inherit;
text-decoration:none;
-js-display:flex;
background-color:transparent;
border-top-style:solid;
border-right-style:solid;
border-bottom-style:solid;
border-left-style:solid;
border-top-width:0px;
border-right-width:0px;
border-bottom-width:0px;
border-left-width:0px;
box-sizing:border-box;
margin-left:0px;
}
.rp_ViewReset
{
-webkit-flex-shrink:0px;
-ms-flex-negative:0px;
flex-shrink:0;
}
.rp_Text
{
padding-top:0px;
color:inherit;
font:inherit;
margin-top:0px;
margin-right:0px;
margin-bottom:0px;
margin-left:0px;
display:inline;
padding-right:0px;
padding-bottom:0px;
padding-left:0px;
text-decoration:none;
white-space:pre-wrap;
word-wrap:break-word;
}
.rp_Image
{
border-top-width:0px;
border-right-width:0px;
border-bottom-width:0px;
border-left-width:0px;
height:auto;
max-height:100%;
max-width:100%;
opacity:0;
}
"
`;
